File: DSA/quickSort.c
#include<stdio.h>
#include<stdlib.h>

int Quick(int* arr , int beg , int end){
    int left , right , temp , loc ; 
    left = beg , right = end,loc=left; 
    A: while (right!=loc && arr[right]<=arr[loc])
    right-- ; 
    if(right==loc)return loc ;
    else{
      temp = arr[loc];
      arr[loc]=arr[right];
      arr[right]=temp ; 
      loc = right ; 
    }
    while (left!=loc && arr[left]>=arr[loc])
    left++;
    if(left==loc)return loc ; 
    else{
        temp = arr[loc];
        arr[loc]=arr[left];
        arr[left]=temp ;
        loc = left ; 
    }
    goto A ;
}

void QuickSort(int* arr , int beg , int end){
    int left,right,loc,top,upper[10],lower[10]; 
    top = -1 ; 
    lower[++top]=beg ; // 0
    upper[top]=end ;   // 8
    while(top!=-1){
    left = lower[top]; //0
    right = upper[top]; //0
    top--; // -1
    loc=Quick(arr , left , right );
    if(left<loc-1){
        lower[++top]= left , upper[top]=loc-1; 
    }
    if(right>loc+1){
        lower[++top]=loc+1 , upper[top]=right ;
    }
    
    }
}

void display(int* arr , int beg , int end ){
    for(int i=beg ; i <=end ; i++){
        printf("Value:%d\n" ,arr[i]);
    }
}

int main(void){
    int arr[]={1,2,3,4,5,6,7,8,9};
    QuickSort(arr , 0 , 8 );
    display(arr , 0 , 8);
}

File: Python/QuickSort.py
def quickLeft(Array , left , right ): 
	print("Start:Left")
	print(Array)
	pivot = left 
	while right!=pivot and Array[right]>=Array[pivot]:
		right -= 1
	if right == pivot : 
		return pivot
	Array[right],Array[pivot] = Array[pivot],Array[right]
	print("End:Left")
	print(Array)
	return quickRight(Array,left,right)
	
def quickRight(Array , left , right) : 
	print("Start:Right")
	print(Array)
	pivot = right
	while left!=pivot and Array[left]<=Array[pivot]:
		left += 1 
	if left == pivot :
		return pivot 
	Array[left],Array[pivot]=Array[pivot],Array[left]
	print("End:Right")
	print(Array)
	return quickLeft(Array,left,right)

def quickSort(Array,beg,end):
	
	if beg>=end:
		return Array 
	pivot = quickLeft(Array , beg , end )
	print("Left Half")
	quickSort(Array , beg , pivot-1)
	print("Right Half")
	quickSort(Array , pivot+1 , end)
	return Array 







Array = [10,9,8,7,5,4,3,2,1]
quickSort(Array,0,len(Array)-1)
print(Array)

